### PR TITLE
Added iOS Objective C Gesture Recognizer Tests

### PR DESCRIPTION
**Pull request by priankakariatyml**

### CODE REVIEW

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+// limitations under the License.
+
+#import <XCTest/XCTest.h>
+
+#import "mediapipe/tasks/ios/common/sources/MPPCommon.h"
+#import "mediapipe/tasks/ios/test/vision/gesture_recognizer/utils/sources/MPPGestureRecognizerResult+ProtobufHelpers.h"
+#import "mediapipe/tasks/ios/test/vision/utils/sources/MPPImage+TestUtils.h"
+#import "mediapipe/tasks/ios/vision/gesture_recognizer/sources/MPPGestureRecognizer.h"
+
+static NSDictionary *const kGestureRecognizerBundleAssetFile =
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m:22]: My last reviewer asked me to specify the generics for all theses files. You should see the changes soon in the FaceLandmarker that is going to GitHub right now.

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+static NSString *const kExpectedErrorDomain = @"com.google.mediapipe.tasks";
+static const float kLandmarksErrorTolerance = 0.03f;
+
+#define AssertEqualErrors(error, expectedError)                                               \
+  XCTAssertNotNil(error);                                                                     \
+  XCTAssertEqualObjects(error.domain, expectedError.domain);                                  \
+  XCTAssertEqual(error.code, expectedError.code);                                             \
+  XCTAssertNotEqual(                                                                          \
+      [error.localizedDescription rangeOfString:expectedError.localizedDescription].location, \
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m:52]: Please use object equality.

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  NSString *filePath = [MPPGestureRecognizerTests filePathWithFileInfo:kExpectedFistLandmarksFile];
+
+  return [MPPGestureRecognizerResult
+      gestureRecognizerResultsFromTextEncodedProtobufFileWithName:filePath
+                                                     gestureLabel:gestureLabel
+                                            shouldRemoveZPosition:YES];
+}
+
+- (void)assertMultiHandLandmarks:(NSArray<NSArray<MPPNormalizedLandmark *> *> *)multiHandLandmarks
+    isApproximatelyEqualToExpectedMultiHandLandmarks:
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m:108]: s/is/are ?

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+  XCTAssertEqual(topHandLandmarks.count, expectedTopHandLandmarks.count);
+  for (int i = 0; i < expectedTopHandLandmarks.count; i++) {
+    MPPNormalizedLandmark *landmark = topHandLandmarks[i];
+    XCTAssertNotNil(landmark);
+    AssertApproximatelyEqualLandmarks(landmark, expectedTopHandLandmarks[i], 0, i);
+  }
+}
+
+- (void)assertMultiHandWorldLandmarks:(NSArray<NSArray<MPPLandmark *> *> *)multiHandWorldLandmarks
+    isApproximatelyEqualToExpectedMultiHandWorldLandmarks:
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m:127]: s/is/are ?

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+  XCTAssertEqual(topHandWorldLandmarks.count, expectedTopHandWorldLandmarks.count);
+  for (int i = 0; i < expectedTopHandWorldLandmarks.count; i++) {
+    MPPLandmark *landmark = topHandWorldLandmarks[i];
+    XCTAssertNotNil(landmark);
+    AssertApproximatelyEqualLandmarks(landmark, expectedTopHandWorldLandmarks[i], 0, i);
+  }
+}
+
+- (void)assertMultiHandGestures:(NSArray<NSArray<MPPCategory *> *> *)multiHandGestures
+    isApproximatelyEqualToExpectedMultiHandGestures:
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m:146]: Same :)

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+  XCTAssertEqual(topHandGestures.count, expectedTopHandGestures.count);
+  for (int i = 0; i < expectedTopHandGestures.count; i++) {
+    MPPCategory *gesture = topHandGestures[i];
+    XCTAssertNotNil(gesture);
+    AssertEqualGestures(gesture, expectedTopHandGestures[i], 0, i);
+  }
+}
+
+- (void)assertGestureRecognizerResult:(MPPGestureRecognizerResult *)gestureRecognizerResult
+    isApproximatelyEqualToExpectedResult:
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m:165]: And last one.

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+  MPPGestureRecognizer *gestureRecognizer =
+      [[MPPGestureRecognizer alloc] initWithOptions:gestureRecognizerOptions error:&error];
+
+  XCTAssertNil(gestureRecognizer);
+  AssertEqualErrors(error, expectedError);
+}
+
+#pragma mark Assert Gesture Recognizer Results
+
+- (MPPImage *)imageWithFileInfo:(NSDictionary *)fileInfo {
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m:224]: FWIW, we should eventually move all the shared code to a helpers file.

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/utils/sources/MPPGestureRecognizerResult+ProtobufHelpers.mm [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                                          shouldRemoveZPosition:(BOOL)removeZPosition {
+  LandmarksDetectionResultProto landmarkDetectionResultProto;
+
+  if (!get_proto_from_pbtxt(fileName.cppString, landmarkDetectionResultProto).ok()) {
+    return nil;
+  }
+
+  if (removeZPosition) {
+    // Remove z position of landmarks, because they are not used in correctness
+    // testing. For video or live stream mode, the z positions varies a lot during
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/test/vision/gesture_recognizer/utils/sources/MPPGestureRecognizerResult+ProtobufHelpers.mm:46]: This comment looks like it uses a shorter line length.

File: mediapipe/tasks/ios/vision/gesture_recognizer/utils/sources/MPPGestureRecognizerResult+Helpers.mm [hunk 9]
```diff
@@ -42,48 +38,33 @@ + (MPPGestureRecognizerResult *)emptyGestureRecognizerResultWithTimestampInMilli
...
-      !worldLandmarksPacket.ValidateAsType<std::vector<LandmarkListProto>>().ok()) {
-    return [MPPGestureRecognizerResult
-        emptyGestureRecognizerResultWithTimestampInMilliseconds:timestampInMilliseconds];
-  }
-
-  const std::vector<ClassificationListProto> &handGesturesClassificationListProtos =
-      handGesturesPacket.Get<std::vector<ClassificationListProto>>();
+    gestureRecognizerResultWithHandGesturesProto:
+        (const std::vector<ClassificationListProto> &)handGesturesProto
+                                  handednessroto:
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/vision/gesture_recognizer/utils/sources/MPPGestureRecognizerResult+Helpers.mm:43]: Missing "O"

File: mediapipe/tasks/ios/vision/gesture_recognizer/utils/sources/MPPGestureRecognizerResult+Helpers.mm [hunk 10]
```diff
@@ -42,48 +38,33 @@ + (MPPGestureRecognizerResult *)emptyGestureRecognizerResultWithTimestampInMilli
...
+                                      (const std::vector<ClassificationListProto> &)handednessProto
+                             handLandmarksPacket:(const std::vector<NormalizedLandmarkListProto> &)
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/ios/vision/gesture_recognizer/utils/sources/MPPGestureRecognizerResult+Helpers.mm:45]: s/Packet/Proto (and below)

**Review by schmidt-sebastian — COMMENTED** [review]
Thanks! Some small changes.

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m (hunk 4 above)
Comment (priankakariatyml) [reply mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m]: grammatically correct. Fixed it.

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m (hunk 1 above)
Comment (priankakariatyml) [reply mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m:27]: Used generics for now. Might update these to use a class `FileInfo`  in future, when I move file path methods to helpers.

File: mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m (hunk 7 above)
Comment (priankakariatyml) [reply mediapipe/tasks/ios/test/vision/gesture_recognizer/MPPGestureRecognizerTests.m]: Will do while writing tests for the image segmenter and slowly migrate all tests.

**Review by schmidt-sebastian — APPROVED** [review]

**Pull request merged**